### PR TITLE
Ported conv2D and added conv2DTranpose 

### PR DESCRIPTION
**Pull request by HemangRajvanshy**
Changes for #72  (still WIP)

Can someone verify that this is the new format for the tf primitives and that I added the arguments correctly? 

Also quick question, how does the evolutionary process differentiate between "inputs" and "args", i.e. could args potentially be passed in as inputs? 

Test Plan:
`python main.py -p problem_multiGaussian -n "test" -s 9 -t -d`
Runs successfully.
More testing suggestions welcome.

### CODE REVIEW

File: codes/block_definitions/utilities/operators_TensorFlow_Keras.py [hunk 1]
```diff
@@ -11,31 +11,63 @@
...
 import numpy as np
 
-### sys relative to root dir
+# sys relative to root dir
 import sys
 from os.path import dirname, realpath
 sys.path.append(dirname(dirname(dirname(dirname(realpath(__file__))))))
 
-### abosulte imports wrt root
-from codes.block_definitions.utilities import argument_types
```
Comment (roddtalebi) [inline codes/block_definitions/utilities/operators_TensorFlow_Keras.py:24]: I would keep these 'imports' here after we add the current directory to the sys.path

almost every single one of our py scripts are like this: import external modules, add root dir to sys.path, import local py scripts relative to root dir.

technically once one py script adds root dir to the sys.path then none of the others need to and then it doesn't matter where (ie could be before or could be after we alter sys.path) we import the local scripts, but we keep it like this for consistency

File: codes/block_definitions/utilities/operators_TensorFlow_Keras.py [hunk 2]
```diff
@@ -11,31 +11,63 @@
...
+                                  padding="same",
+                                  activation=activation,
+                                  data_format="channels_last"
+                                  )(input_tensor)
+
+
+operator_dict[conv2D_layer] = {"inputs": [tf.keras.layers],
+                               "output": tf.keras.layers,
+                               "args": [argument_types.ArgumentType_Pow2, argument_types.ArgumentType_TFFilterSize, argument_types.ArgumentType_TFActivation]
+                               }
```
Comment (roddtalebi) [inline codes/block_definitions/utilities/operators_TensorFlow_Keras.py:53]: conv2D_layers looks good

File: codes/block_definitions/utilities/operators_TensorFlow_Keras.py [hunk 3]
```diff
@@ -11,31 +11,63 @@
...
+                                           kernel_size=kernel_size,
+                                           padding="same",
+                                           activation=activation,
+                                           data_format="channels_last"
+                                           )(input_tensor)
 
 
+operator_dict[conv2DTranspose_layer] = {"inputs": [tf.keras.layers],
+                                        "output": tf.keras.layers,
+                                        "args": [argument_types.ArgumentType_Pow2, argument_types.ArgumentType_TFFilterSize, argument_types.ArgumentType_TFActivation]
```
Comment (roddtalebi) [inline codes/block_definitions/utilities/operators_TensorFlow_Keras.py:72]: conv2DTranspose_layer also looks good

**Review by roddtalebi — COMMENTED** [review]
> Can someone verify that this is the new format for the tf primitives and that I added the arguments correctly?

Looks good

> Also quick question, how does the evolutionary process differentiate between "inputs" and "args", i.e. could args potentially be passed in as inputs?

this is where the 'operator_dict' comes in...

(note: remember that our genome has 'input', 'main', and 'output' nodes where only 'main' nodes have primitives/operators, and primitives and operators are the same thing)
Say we are initializing a genome...right off the bat, I know the datatypes of my 'input' nodes. Then I go to the first 'main' node (index 0), and I want to find an operator who's expected input datatypes are included in the list of data types given by the input nodes...ie say ```previous_node_datatypes``` is a list of data types given by the 'input' nodes, and ```expected_datatypes = operator_dict[my_ftn]["inputs"]``` where ```my_ftn``` was randomly chosen from all available primitives/operators, then I would fill the first 'main' node with primitive ```my_ftn``` if all the datatypes in expected_datatypes are in previous_node_datatypes; if that is not the case, then I randomly select another primitive/operator until that is satisfied.

Soooo, if ```operator_dict[my_ftn]["inputs"]``` returns something we would consider an 'argument' data type, under the assumption that ```previous_node_datatypes``` would never contain any 'argument' data type, then that operator my_ftn will NEVER get used in the genome. So, No, we can't pass an argument datatype into 'inputs' key/value pair of the operator dict

**Comment by HemangRajvanshy** [comment]
Added `Conv3D` and `Conv3DTranspose` primitives. Left out one-dimensional convolutions for now as I feel they might be redundant with 2D convolutions and not highly applicable for image processing. 

PR ready for review and merge

**Review by roddtalebi — APPROVED** [review]

**Pull request merged**